### PR TITLE
Implement multipart upload for GCS artifact store

### PR DESCRIPTION
**Pull request by gabrielfu**
[Open in GitHub Codespaces](https://codespaces.new/gabrielfu/mlflow/pull/10124?quickstart=1)

### Related Issues/PRs

 #9521

### What changes are proposed in this pull request?

Implement MPU for GCS.

Python GCS client provides only one single function [upload_chunks_concurrently](https://cloud.google.com/python/docs/reference/storage/latest/google.cloud.storage.transfer_manager#google_cloud_storage_transfer_manager_upload_chunks_concurrently) which is unusable behind proxy & if we want presigned url. Thus, we need to copy some code from that function and use the low level XML api.

We require `google-cloud-storage>=2.12.0` to support some operations.

Example script:
```python
import math
import os
import requests

os.environ["GOOGLE_APPLICATION_CREDENTIALS"] = "/path/to/service_account.json"

from mlflow.entities.multipart_upload import MultipartUploadPart
from mlflow.store.artifact.gcs_artifact_repo import GCSArtifactRepository

local_file = "my-file"
bucket_name = "mlflow-mpu-test"
artifact_path = "my-folder"

repo = GCSArtifactRepository(f"gs://{bucket_name}")
chunk_size = 5 * 1024 * 1024
file_size = os.path.getsize(local_file)
num_parts = math.ceil(file_size / chunk_size)
create = repo.create_multipart_upload(local_file, num_parts , artifact_path)
parts = []
try:
    for i, credential in enumerate(create.credentials):
        print(f"Uploading part #{credential.part_number}")
        with open(local_file, "rb") as f:
            f.seek(i * chunk_size)
            chunk = f.read(chunk_size)

        response = requests.put(
            credential.url,
            data=chunk,
        )
        response.raise_for_status()
        parts.append(
            MultipartUploadPart(
                part_number=credential.part_number,
                etag=response.headers["ETag"],
            )
        )

    print("Uploaded all parts:", parts)
    repo.complete_multipart_upload(local_file, create.upload_id, parts, artifact_path)
    print("Multipart upload completed")
except Exception as e:
    print("Error uploading parts, aborting upload:", e)
    repo.abort_multipart_upload(local_file, create.upload_id, artifact_path)
```

### How is this PR tested?

- [ ] Existing unit/integration tests
- [X] New unit/integration tests
- [ ] Manual tests

### Does this PR require documentation update?

- [X] No. You can skip the rest of this section.
- [ ] Yes. I've updated:
  - [ ] Examples
  - [ ] API references
  - [ ] Instructions

#### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

Implements multipart upload for Google Cloud Storage artifact store.

#### What component(s), interfaces, languages, and integrations does this PR affect?

Components

- [X] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/gateway`: AI Gateway service, Gateway client APIs, third-party Gateway integrations
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [ ] `area/recipes`: Recipes, Recipe APIs, Recipe configs, Recipe Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface

- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language

- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations

- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<a name="release-note-category"></a>

#### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [X] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

### CODE REVIEW

File: setup.py [hunk 1]
```diff
@@ -147,7 +147,7 @@ def run(self):
             # Required to log artifacts and models to AWS S3 artifact locations
             "boto3",
             # Required to log artifacts and models to GCS artifact locations
-            "google-cloud-storage>=1.30.0",
+            "google-cloud-storage>=2.12.0",
```
Comment (harupy) [inline setup.py:150]: Is < 2.12 missing required features?

Comment (gabrielfu) [reply setup.py:150]: - `XMLMPUContainer` class from `google.resumable_media.requests` [dependency](https://github.com/googleapis/google-resumable-media-python/blob/v2.6.0/google/resumable_media/requests/upload.py#L560). Without this, we need to build the xml request body ourselves (e.g. [this](https://cloud.google.com/storage/docs/xml-api/post-object-complete)).
- [some headers added](https://github.com/googleapis/python-storage/blob/v2.12.0/google/cloud/storage/_helpers.py#L618-L621) on mpu create request

Comment (gabrielfu) [reply setup.py:150]: mpu support is quite new on the python gcs client

Comment (harupy) [reply setup.py:150]: Got it.

> Without this, we need to build the xml request body ourselves

This sounds painful.

Comment (harupy) [reply setup.py:150]: Is it possible to enable mpu for GCS only when `XMLMPUContainer` is importable?

Comment (gabrielfu) [reply setup.py:150]: that's a good idea, let me think how to make it clean

File: mlflow/exceptions.py [hunk 2]
```diff
@@ -143,4 +143,7 @@ class InvalidUrlException(MlflowException):
 class UnsupportedMultipartUploadException(MlflowException):
     """Exception thrown when multipart upload is unsupported by an artifact repository"""
 
-    pass
+    message = "Multipart upload is not supported for the current artifact repository"
+
+    def __init__(self):
+        super().__init__(self.message, error_code=NOT_IMPLEMENTED)
```
Comment (gabrielfu) [inline mlflow/exceptions.py:149]: changed this to 501 response which should make more sense. Also, removing the artifact repo class name from the message.

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 3]
```diff
@@ -141,3 +148,109 @@ def delete_artifacts(self, artifact_path=None):
         blobs = gcs_bucket.list_blobs(prefix=f"{dest_path}")
         for blob in blobs:
             blob.delete()
+
+    @staticmethod
+    def _validate_support_mpu():
+        if not compare_package_version(
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:154]: Does `compare_package_version` throw if `google-cloud-storage` isn't installed?

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 4]
```diff
@@ -141,3 +148,109 @@ def delete_artifacts(self, artifact_path=None):
...
+            "google-cloud-storage", "2.12.0", ">="
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:155]: Does `google-cloud-storage` have `__version__`?

File: mlflow/utils/version.py [hunk 5]
```diff
@@ -0,0 +1,28 @@
...
+    :param package: The name of the package.
+    :param version: The version to compare with.
+    :param cmp: The comparison operator.
+    :return: True if the version of the package satisfies the comparison.
+    """
+
+    if cmp not in operators:
+        raise ValueError(f"Invalid comparison operator: {cmp}")
+    op = operators[cmp]
+    return op(Version(importlib.metadata.version(package)), Version(version))
```
Comment (harupy) [inline mlflow/utils/version.py:28]: I'd make this function return a `Version` instance.

File: mlflow/store/artifact/gcs_artifact_repo.py (hunk 3 above)
Comment (gabrielfu) [reply mlflow/store/artifact/gcs_artifact_repo.py:154]: yes but `GCSArtifactRepository` won't be able to be instantiated in this case

File: mlflow/utils/version.py (hunk 5 above)
Comment (gabrielfu) [reply mlflow/utils/version.py:28]: in this case, the caller will need to import `Version` and to compare with the target version themselves?

File: mlflow/store/artifact/gcs_artifact_repo.py (hunk 4 above)
Comment (gabrielfu) [reply mlflow/store/artifact/gcs_artifact_repo.py:155]: iiuc, `importlib.metadata.version` is the quickest way to get the version of a package as it doesn't need to perform the import action

Comment (harupy) [reply mlflow/store/artifact/gcs_artifact_repo.py:155]: I saw Flask 3.0 drop `__version__` last week and recommend `importlib.metadata.version`. Let's use it.

File: mlflow/utils/version.py (hunk 5 above)
Comment (gabrielfu) [reply mlflow/utils/version.py:28]: maybe i'll just remove this function and inline in gcs repo, what do you think?

Comment (harupy) [reply mlflow/utils/version.py:28]: >  i'll just remove this function and inline in gcs repo

Can you share what the diff would look like?

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 6]
```diff
@@ -141,3 +148,109 @@ def delete_artifacts(self, artifact_path=None):
...
+        ) or not compare_package_version("google-resumable-media", "2.6.0", ">="):
```
Comment (gabrielfu) [inline mlflow/store/artifact/gcs_artifact_repo.py:156]:
```suggestion
        if (
                Version(importlib.metadata.version("google-cloud-storage")) < Version("2.12.0")
                or Version(importlib.metadata.version("google-resumable-media")) < Version("2.6.0")
        ):
```

@harupy

File: mlflow/utils/version.py (hunk 5 above)
Comment (harupy) [reply mlflow/utils/version.py:28]: https://github.com/mlflow/mlflow/pull/10124#discussion_r1377258518 looks good to me.

File: mlflow/exceptions.py (hunk 2 above)
Comment (harupy) [reply mlflow/exceptions.py:149]: sg!

File: mlflow/store/artifact/gcs_artifact_repo.py [hunk 7]
```diff
@@ -141,3 +150,109 @@ def delete_artifacts(self, artifact_path=None):
...
+        )
+        headers = {**base_headers, **_headers_from_metadata(object_metadata)}
+
+        if blob.user_project is not None:
+            headers["x-goog-user-project"] = blob.user_project
+
+        if blob.kms_key_name is not None and "cryptoKeyVersions" not in blob.kms_key_name:
+            headers["x-goog-encryption-kms-key-name"] = blob.kms_key_name
+
+        return transport, url, headers, content_type
```
Comment (harupy) [inline mlflow/store/artifact/gcs_artifact_repo.py:184]: Can we create a named tuple for this?

File: setup.py [hunk 8]
```diff
@@ -167,7 +167,7 @@ def run(self):
         "databricks": [
             # Required to write model artifacts to unity catalog locations
             "azure-storage-file-datalake>12",
-            "google-cloud-storage>=1.30.0",
+            "google-cloud-storage>=2.12.0",
```
Comment (harupy) [inline setup.py:170]: I guess we can revert this now?

File: mlflow/exceptions.py [hunk 9]
```diff
@@ -143,4 +143,7 @@ class InvalidUrlException(MlflowException):
 class UnsupportedMultipartUploadException(MlflowException):
```
Comment (harupy) [inline mlflow/exceptions.py:143]: Can we mark this as private to avoid accidentally documenting this class?

File: mlflow/exceptions.py [hunk 10]
```diff
@@ -143,4 +143,7 @@ class InvalidUrlException(MlflowException):
...
     """Exception thrown when multipart upload is unsupported by an artifact repository"""
 
-    pass
+    message = "Multipart upload is not supported for the current artifact repository"
```
Comment (harupy) [inline mlflow/exceptions.py:146]:
```suggestion
    MESSAGE = "Multipart upload is not supported for the current artifact repository"
```

File: mlflow/store/artifact/gcs_artifact_repo.py (hunk 7 above)
Comment (harupy) [reply mlflow/store/artifact/gcs_artifact_repo.py:184]: Unpacking like this is error-prone:

```
a, b, c, d = ...
```

**Review by harupy — APPROVED** [review]
left a few more comments, otherwise LGTM!

File: mlflow/store/artifact/gcs_artifact_repo.py (hunk 7 above)
Comment (gabrielfu) [reply mlflow/store/artifact/gcs_artifact_repo.py]: sure

File: setup.py (hunk 8 above)
Comment (gabrielfu) [reply setup.py]: yes i forgot about it

File: mlflow/exceptions.py (hunk 9 above)
Comment (gabrielfu) [reply mlflow/exceptions.py]: sure

**Pull request merged**